pkg/plugins/manager/loader/finder/fs.go

```diff
@@ -0,0 +1,286 @@
+				JSONData: data,
+				FS:       plugins.NewLocalFS(files, pluginDir),
+			},
+		}
+	}
+
+	var result []*plugins.FoundBundle
+	for dir := range foundPlugins {
+		ancestors := strings.Split(dir, string(filepath.Separator))
+		ancestors = ancestors[0 : len(ancestors)-1]
+
+		pluginPath := ""
+		if runtime.GOOS != "windows" && filepath.IsAbs(dir) {
+			pluginPath = "/"
+		}
+		add := true
+		for _, ancestor := range ancestors {
+			pluginPath = filepath.Join(pluginPath, ancestor)
+			if _, ok := foundPlugins[pluginPath]; ok {
+				if fp, exists := res[pluginPath]; exists {
+					fp.Children = append(fp.Children, &res[dir].Primary)
+					add = false
+					break
+				}
+			}
+		}
+		if add {
+			result = append(result, res[dir])
+		}
+	}
+
+	return result, nil
+}
+
+func (f *FS) getAbsPluginJSONPaths(path string) ([]string, error) {
+	var pluginJSONPaths []string
+
+	var err error
+	path, err = filepath.Abs(path)
+	if err != nil {
+		return []string{}, err
+	}
+
+	if err = walk(path, true, true,
+		func(currentPath string, fi os.FileInfo, err error) error {
+			if err != nil {
+				if errors.Is(err, os.ErrNotExist) {
+					f.log.Error("Couldn't scan directory since it doesn't exist", "pluginDir", path, "err", err)
+					return nil
+				}
+				if errors.Is(err, os.ErrPermission) {
+					f.log.Error("Couldn't scan directory due to lack of permissions", "pluginDir", path, "err", err)
+					return nil
+				}
+
+				return fmt.Errorf("filepath.Walk reported an error for %q: %w", currentPath, err)
+			}
+
+			if fi.Name() == "node_modules" {
+				return util.ErrWalkSkipDir
+			}
+
+			if fi.IsDir() {
+				return nil
+			}
+
+			if fi.Name() != "plugin.json" {
+				return nil
+			}
+
+			pluginJSONPaths = append(pluginJSONPaths, currentPath)
+			return nil
+		}); err != nil {
+		return []string{}, err
+	}
+
+	return pluginJSONPaths, nil
+}
+
+func (f *FS) readPluginJSON(pluginJSONPath string) (plugins.JSONData, error) {
+	f.log.Debug("Loading plugin", "path", pluginJSONPath)
+
+	if !strings.EqualFold(filepath.Ext(pluginJSONPath), ".json") {
+		return plugins.JSONData{}, ErrInvalidPluginJSONFilePath
+	}
+
+	absPluginJSONPath, err := filepath.Abs(pluginJSONPath)
+	if err != nil {
+		return plugins.JSONData{}, err
+	}
+
+	// Wrapping in filepath.Clean to properly handle
+	// gosec G304 Potential file inclusion via variable rule.
+	reader, err := os.Open(filepath.Clean(absPluginJSONPath))
+	if err != nil {
+		return plugins.JSONData{}, err
+	}
+	defer func() {
+		if reader == nil {
+			return
+		}
+		if err = reader.Close(); err != nil {
+			f.log.Warn("Failed to close JSON file", "path", pluginJSONPath, "err", err)
+		}
+	}()
+
+	plugin := plugins.JSONData{}
+	if err = json.NewDecoder(reader).Decode(&plugin); err != nil {
+		return plugins.JSONData{}, err
+	}
+
+	if err = validatePluginJSON(plugin); err != nil {
+		return plugins.JSONData{}, err
+	}
+
+	if plugin.ID == "grafana-piechart-panel" {
+		plugin.Name = "Pie Chart (old)"
+	}
```

Comment:
isn't there a better place to do this? :sweat_smile: I see that this was like this before, but in case this could be done somewhere else, maybe consider leaving a // TODO here?

Reply:
Haha yes this line kills me. Unfortunately, there's no better place for it right now. There was an issue for it https://github.com/grafana/grafana/issues/55184 - maybe something that can be revisited during the angular deprecation stuff perhaps CC @tolzhabayev

Reply:
If this gets merged https://github.com/grafana/grafana/pull/63915/files then I think we can just remove this naming hack